Add llama package

gridbugs opened the pull request:
Llama is a library for declaratively building software-defined synthesizers.

This package uses a small rust library to do low-level operations such as sending samples to the sound card. The -rustup version of the package is identical to the default version except that it lacks the conf-rust-2021 dependency. This is to deal with the fact that I expect rust developers would find it off-putting for an ocaml package to install rust globally on their system, as they likely manage their own rust installation with rustup.

gridbugs commented:
Just realized that I needed to mark the library which depends on `sdl2` as optional because `ocamlsdl2` is a depopt. I'll re-do the release to fix this issue and update this PR with the new hash.

gridbugs commented:
Ok I fixed the issue with `ocamlsdl2` by splitting the package into `llama` and `llama_interactive` the latter of which depends on `ocamlsdl2` and there are no more optional dependencies which simplifies things.

As I mentioned in the commit message I'm releasing a default and -rustup version of both packages so that rust devs won't have to install rust globally on their system if they'd prefer to manage it with rustup. The only difference between llama-0.0.1/opam and llama-0.0.1-rustup/opam is that the former has an additional dependency on `conf-rust-2021`. The files llama_interactive-0.0.1/opam and llama_interactive-0.0.1-rustup/opam are identical but I think I still need both because they depend on `"llama" {= version}`. Please let me know if there's a better way to solve this problem!

gridbugs commented:
The more I think about the less I like my idea of having a default and -rustup version of these packages. I'm thinking of erring on the side of removing the -rustup versions and keeping `conf-rust-2021` among the dependencies. People who use rustup can always pass `--assume-depexts` to avoid installing rust system-wide. Unfortunately `llama_interactive` also depends (transitively) on the depext `sdl2` so I guess people will need to manually run `opam install conf-rust-2021 --assume-depexts` before installing this package. That is `opam install llama_interactive --assume-depexts` won't work as it would also skip the sdl2 deps.

mseri commented:
I had come to leave a comment exactly in the spirit of your last comment. People can install rust manually and assume depexts, so I don't think there is any need to have a duplicate package release

gridbugs commented:
Thanks for confirming that! I have updated this PR to just include the manifests which depend on `conf-rust-2021` and removed the note from the description explaining the -rustup package.

mseri commented:
You need to move the file from `packages/llama/llama-0.0.1/opam` to `packages/llama/llama.0.0.1/opam` otherwise opam is unable to parse the name and the version correctly

gridbugs commented:
> You need to move the file from `packages/llama/llama-0.0.1/opam` to `packages/llama/llama.0.0.1/opam` otherwise opam is unable to parse the name and the version correctly

Done. Thanks for pointing that out!

gridbugs commented:
Ah looks like it fails to link on alpine. I've been testing on macos and ubuntu. Will investigate.

mseri commented:
We can make it unavailable on alpine for the time being

mseri commented on packages/llama/llama.0.0.1/opam:
```diff
@@ -0,0 +1,40 @@
+    "@doc" {with-doc}
+  ]
+]
+dev-repo: "git+https://github.com/gridbugs/llama.git"
```
```suggestion
dev-repo: "git+https://github.com/gridbugs/llama.git"
available: os-family != "alpine"
```

mseri commented:
It is actually failing to link properyly on many distributions:
```
alpine-3.16-ocaml-4.14
[llama.0.0.1 (failed: package `rustix v0.38.2` cannot be built because it requires rustc 1.63 or newer, while the currently active rustc version is 1.60.0)](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/7b68fa19928dbb9d3ab1a94257e3e99f6d2a5c47/variant/distributions,alpine-3.16-ocaml-4.14,llama.0.0.1)
[llama_interactive.0.0.1 (failed: llama.0.0.1 failed to build)](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/7b68fa19928dbb9d3ab1a94257e3e99f6d2a5c47/variant/distributions,alpine-3.16-ocaml-4.14,llama_interactive.0.0.1)
alpine-3.17-ocaml-4.14
[llama_interactive.0.0.1 (failed: Error during linking (exit code 1))](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/7b68fa19928dbb9d3ab1a94257e3e99f6d2a5c47/variant/distributions,alpine-3.17-ocaml-4.14,llama_interactive.0.0.1)
debian-testing-ocaml-4.14
[llama_interactive.0.0.1 (failed: Error during linking (exit code 1))](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/7b68fa19928dbb9d3ab1a94257e3e99f6d2a5c47/variant/distributions,debian-testing-ocaml-4.14,llama_interactive.0.0.1)
debian-unstable-ocaml-4.14
[llama_interactive.0.0.1 (failed: Error during linking (exit code 1))](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/7b68fa19928dbb9d3ab1a94257e3e99f6d2a5c47/variant/distributions,debian-unstable-ocaml-4.14,llama_interactive.0.0.1)
oraclelinux-8-ocaml-4.14
[llama.0.0.1 (failed: conf-alsa.1 failed to build)](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/7b68fa19928dbb9d3ab1a94257e3e99f6d2a5c47/variant/distributions,oraclelinux-8-ocaml-4.14,llama.0.0.1)
[llama_interactive.0.0.1 (failed: System package install failed with exit code 1 at command:)](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/7b68fa19928dbb9d3ab1a94257e3e99f6d2a5c47/variant/distributions,oraclelinux-8-ocaml-4.14,llama_interactive.0.0.1)
ubuntu-18.04-ocaml-4.14
[llama_interactive.0.0.1 (failed: ocamlsdl2.0.04 failed to build)](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/7b68fa19928dbb9d3ab1a94257e3e99f6d2a5c47/variant/distributions,ubuntu-18.04-ocaml-4.14,llama_interactive.0.0.1)
ubuntu-22.10-ocaml-4.14
[llama_interactive.0.0.1 (failed: Error during linking (exit code 1))](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/7b68fa19928dbb9d3ab1a94257e3e99f6d2a5c47/variant/distributions,ubuntu-22.10-ocaml-4.14,llama_interactive.0.0.1)
macos
macos-homebrew-ocaml-4.14-arm64 (experimental)
[llama_interactive.0.0.1 (failed: Error during linking (exit code 1))](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/7b68fa19928dbb9d3ab1a94257e3e99f6d2a5c47/variant/macos,macos-homebrew-ocaml-4.14-arm64,llama_interactive.0.0.1)

gridbugs commented:
oh interesting I've been doing all my development of this package on `macos-homebrew-ocaml-4.14-arm64`

gridbugs commented:
Trying to reproduce the macos failure by running the script at the top of https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/7b68fa19928dbb9d3ab1a94257e3e99f6d2a5c47/variant/macos,macos-homebrew-ocaml-4.14-arm64,llama_interactive.0.0.1 and I get the error:

```
Dockerfile:1
--------------------
   1 | >>> FROM macos-homebrew-ocaml-4.14
   2 |     USER 1000:1000
   3 |     RUN ln -f ~/local/bin/opam-2.1 ~/local/bin/opam
--------------------
ERROR: failed to solve: macos-homebrew-ocaml-4.14: pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed

```

mseri commented:
I think only the linux ones are reproducible, mac is some special thing

gridbugs commented:
The mac issue might be related to the linker not knowing the right path to the SDL library. I can try using pkgconfig to find the right path.

gridbugs commented:
Even on my mac when I run pkgconfig I get a different path from the default though it seems to still build despite this:
```
$ pkg-config --libs sdl2
-L/opt/homebrew/lib -lSDL2
```

mseri commented:
As a rule of thumbs I would recommend using pkgconfig because different installations of homebrew (especially wrt different architectures) or macoports will have different paths usually

mseri commented:
the failure seems more of an issue with ocamlsdl2 though

gridbugs commented:
I replaced ocamlsdl2 with tsdl and I no longer need to specify linker flags for sdl. I also changed to using dune-configurator to generate the sexp files of linker flags needed for the low-level audio library which is a lot cleaner than manually running `pkg-config` from the dune file. Thanks for the suggestions.

I've also ended up splitting the `llama` package into two - `llama_core` which just defines the core data types and operations and has no dependencies besides ocaml and dune, and `llama` which includes logic for playing audio which depends on rust. This will make it possible for people to write libraries of sound effects and filters for llama without needing to depend on rust. Note that the `llama_core` package the source url is for the release of `llama` which doesn't bundle the rust dependencies since this is a much smaller archive. `llama` and `llama_interactive` both still point to the archive which vendors all the rust libraries.

mseri commented on packages/llama_core/llama_core.0.0.1/opam:
```diff
@@ -0,0 +1,37 @@
+homepage: "https://github.com/gridbugs/llama"
+bug-reports: "https://github.com/gridbugs/llama/issues"
+depends: [
+  "dune" {>= "3.0"}
```
```suggestion
  "ocaml" {>= "4.13"}
  "dune" {>= "3.0"}
```
Due to
```
#=== ERROR while compiling llama_core.0.0.1 ===================================#
# context              2.2.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.4.02.3 | pinned(https://github.com/gridbugs/llama/archive/refs/tags/0.0.1.tar.gz)
# path                 ~/.opam/4.02/.opam-switch/build/llama_core.0.0.1
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p llama_core -j 31 @install
# exit-code            1
# env-file             ~/.opam/log/llama_core-7-941b83.env
# output-file          ~/.opam/log/llama_core-7-941b83.out
### output ###
# (cd _build/default && /home/opam/.opam/4.02/bin/ocamldep.opt -modules -impl examples/midi/llama_midi.ml) > _build/default/examples/midi/.llama_midi.objs/llama_midi.impl.d
# File "examples/midi/llama_midi.ml", line 26, characters 9-13:
# Error: Syntax error
# (cd _build/default && /home/opam/.opam/4.02/bin/ocamldep.opt -modules -intf examples/midi/byte_array_parser.mli) > _build/default/examples/midi/.llama_midi.objs/llama_midi__Byte_array_parser.intf.d
# File "examples/midi/byte_array_parser.mli", line 10, characters 6-9:
# Error: Syntax error: operator expected.
# (cd _build/default && /home/opam/.opam/4.02/bin/ocamldep.opt -modules -impl examples/midi/byte_array_parser.ml) > _build/default/examples/midi/.llama_midi.objs/llama_midi__Byte_array_parser.impl.d
# File "examples/midi/byte_array_parser.ml", line 20, characters 6-9:
# Error: Syntax error: operator expected.
# (cd _build/default && /home/opam/.opam/4.02/bin/ocamlc.opt -w -40 -g -bin-annot -I examples/midi/.llama_midi.objs/byte -no-alias-deps -open Llama_midi__ -o examples/midi/.llama_midi.objs/byte/llama_midi.cmi -c -intf examples/midi/llama_midi.mli)
# File "examples/midi/llama_midi.mli", line 23, characters 41-76:
# Error: Unbound type constructor result
# (cd _build/default && /home/opam/.opam/4.02/bin/ocamlc.opt -w -40 -g -bin-annot -I src/core/.llama_core.objs/byte -no-alias-deps -open Llama_core__ -o src/core/.llama_core.objs/byte/llama_core__Float.cmi -c -intf src/core/float.mli)
# File "src/core/float.mli", line 2, characters 10-22:
# Error: Unbound module Stdlib
# (cd _build/default && /home/opam/.opam/4.02/bin/ocamlc.opt -w -40 -g -bin-annot -I src/core/.llama_core.objs/byte -no-alias-deps -open Llama_core__ -o src/core/.llama_core.objs/byte/llama_core__Array.cmi -c -intf src/core/array.mli)
# File "src/core/array.mli", line 2, characters 10-28:
# Error: Unbound module Stdlib
# (cd _build/default && /home/opam/.opam/4.02/bin/ocamlc.opt -w -40 -g -bin-annot -I src/core/.llama_core.objs/byte -no-alias-deps -open Llama_core__ -o src/core/.llama_core.objs/byte/llama_core__List.cmi -c -intf src/core/list.mli)
# File "src/core/list.mli", line 2, characters 10-27:
# Error: Unbound module Stdlib
```

```
#=== ERROR while compiling llama_core.0.0.1 ===================================#
# context              2.2.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.4.12.1 | pinned(https://github.com/gridbugs/llama/archive/refs/tags/0.0.1.tar.gz)
# path                 ~/.opam/4.12/.opam-switch/build/llama_core.0.0.1
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p llama_core -j 31 @install
# exit-code            1
# env-file             ~/.opam/log/llama_core-7-4719d4.env
# output-file          ~/.opam/log/llama_core-7-4719d4.out
### output ###
# (cd _build/default && /home/opam/.opam/4.12/bin/ocamlc.opt -w -40 -g -bin-annot -I src/core/.llama_core.objs/byte -intf-suffix .ml -no-alias-deps -open Llama_core__ -o src/core/.llama_core.objs/byte/llama_core__Modules.cmo -c -impl src/core/modules.ml)
# File "src/core/modules.ml", line 348, characters 22-29:
# 348 |           let index = Int.min (index + 1) (Array.length t.data) in
#                             ^^^^^^^
# Error: Unbound value Int.min
# (cd _build/default && /home/opam/.opam/4.12/bin/ocamlopt.opt -w -40 -g -I src/core/.llama_core.objs/byte -I src/core/.llama_core.objs/native -intf-suffix .ml -no-alias-deps -open Llama_core__ -o src/core/.llama_core.objs/native/llama_core__Modules.cmx -c -impl src/core/modules.ml)
# File "src/core/modules.ml", line 348, characters 22-29:
# 348 |           let index = Int.min (index + 1) (Array.length t.data) in
#                             ^^^^^^^
# Error: Unbound value Int.min
```

mseri commented:
Fantastic! Thanks. I added a small fix on `llama_core` dependencies, can you upstream it? There is no need to re-release or re-tag from your side, opam packages are compiled using the metadata from the opam repository releases

mseri commented:
This is now building fine in all systems where it can. Thanks a lot.

Please consider announcing the release on discuss.ocaml.org. We have a community category and an announcement tag for this purpose

gridbugs commented:
> Please consider announcing the release on discuss.ocaml.org. We have a community category and an announcement tag for this purpose

Will do. Thanks for your help with the release.
